mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusContract.java

```diff
@@ -0,0 +1,266 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Mockito.when;
+
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.TestId;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.ImmutableSet;
+
+public interface EventBusContract {
+    MailboxListener.MailboxEvent event = mock(MailboxListener.MailboxEvent.class);
+
+    class GroupA extends Group {}
+    class GroupB extends Group {}
+
+    MailboxId ID_1 = TestId.of(18);
+    MailboxId ID_2 = TestId.of(24);
+    ImmutableSet<RegistrationKey> NO_KEYS = ImmutableSet.of();
+    MailboxIdRegistrationKey KEY_1 = new MailboxIdRegistrationKey(ID_1);
+    MailboxIdRegistrationKey KEY_2 = new MailboxIdRegistrationKey(ID_2);
+
+    EventBus eventBus();
+
+    default MailboxListener newListener() {
+        MailboxListener listener = mock(MailboxListener.class);
+        when(listener.getExecutionMode()).thenReturn(MailboxListener.ExecutionMode.SYNCHRONOUS);
+        return listener;
+    }
+
+    @Test
+    default void listenerGroupShouldReceiveEvents() {
+        MailboxListener listener = newListener();
+
+        eventBus().register(listener, new GroupA());
+
+        eventBus().dispatch(event, NO_KEYS).block();
+
+        verify(listener, times(1)).event(any());
+    }
+
+    @Test
+    default void eachListenerGroupShouldReceiveEvents() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+        eventBus().register(listener, new GroupA());
+        eventBus().register(listener2, new GroupB());
+
+        eventBus().dispatch(event, NO_KEYS).block();
+
+        verify(listener, times(1)).event(any());
+        verify(listener2, times(1)).event(any());
+    }
+
+    @Test
+    default void unregisteredGroupListenerShouldNotReceiveEvents() {
+        MailboxListener listener = newListener();
+        Registration registration = eventBus().register(listener, new GroupA());
+
+        registration.unregister();
+
+        eventBus().dispatch(event, NO_KEYS).block();
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void registerShouldThrowWhenAGroupIsAlreadyUsed() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+
+        eventBus().register(listener, new GroupA());
+
+        assertThatThrownBy(() -> eventBus().register(listener2, new GroupA()))
+            .isInstanceOf(GroupAlreadyRegistered.class);
+    }
+
+    @Test
+    default void registerShouldNotThrowOnAnUnregisteredGroup() {
```

Comment:
this test only verify re-register after unregister does not throw exception. 
missing `registerShouldReceiveEventOnAnUnregisteredGroup` to verify re-register work well(it receives an event). IMO

Reply:
I can add the test you describe